### PR TITLE
Rewrite right distributed table in local join.

### PR DESCRIPTION
**Pull request by abel-cheng**
Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Rewrite right distributed table in local join.
solves #25809

### CODE REVIEW

**Review by vdimir — COMMENTED** [review]
Code LGTM. 

Let's also add test with case from linked issue (and maybe other ones with different `distributed_product_mode` values and `JOIN` ? )

**Comment by abel-cheng** [comment]
@vdimir 
Thx for your advice. 
The linked issue is found on a clickhouse cluster with more than one node, which test case can I add under this condition? Can you show me some examples?

**Comment by vdimir** [comment]
@abel-cheng 

In the config for stateless tests some clusters on localhost are defined https://github.com/ClickHouse/ClickHouse/blob/68a3baf3cd9b74f8aefc770b9e2ab52eccf22dca/tests/config/config.d/clusters.xml

It used with `Distributed` table engine, for example in https://github.com/ClickHouse/ClickHouse/blob/68a3baf3cd9b74f8aefc770b9e2ab52eccf22dca/tests/queries/0_stateless/01901_in_literal_shard_prune.sql and other tests (or you can find on other examples searching by cluster name in `tests/`)

For more complex use-cases integration tests can be used https://github.com/ClickHouse/ClickHouse/blob/68a3baf3cd9b74f8aefc770b9e2ab52eccf22dca/tests/integration/test_cluster_all_replicas/test.py

**Comment by abel-cheng** [comment]
@vdimir 
Thanks for your advice, the tests is added now.

File: src/Interpreters/InJoinSubqueriesPreprocessor.cpp [hunk 1]
```diff
@@ -191,13 +191,25 @@ class NonGlobalSubqueryMatcher
...
+                    if (!renamed.empty()) //-V547
+                        data.renamed_tables.emplace_back(subquery, std::move(renamed));
+                }
+                else if (table->database_and_table_name)
+                {
+                    auto tb = node.table_expression;
+                    std::vector<ASTPtr> renamed;
+                    NonGlobalTableVisitor::Data table_data{data.getContext(), data.checker, renamed, nullptr, table_join};
+                    NonGlobalTableVisitor(table_data).visit(tb);
+                    if (!renamed.empty())
```
Comment (vdimir) [inline src/Interpreters/InJoinSubqueriesPreprocessor.cpp:210]:
```suggestion
                    if (!renamed.empty()) //-V547
```

File: tests/queries/0_stateless/02115_rewrite_local_join_right_distribute_table.sql [hunk 2]
```diff
@@ -0,0 +1,32 @@
+CREATE DATABASE IF NOT EXISTS test_02115;
```
Comment (vdimir) [inline tests/queries/0_stateless/02115_rewrite_local_join_right_distribute_table.sql:1]:
```suggestion
-- Tags: global, no-parallel
CREATE DATABASE IF NOT EXISTS test_02115;
```

File: tests/queries/0_stateless/02115_rewrite_local_join_right_distribute_table.sql [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+set distributed_product_mode = 'global';
+select * from t1_all t1 where t1.a in (select t2.a from t2_all t2);
+explain syntax select t1.* from t1_all t1 join t2_all t2 on t1.a = t2.a;
+select t1.* from t1_all t1 join t2_all t2 on t1.a = t2.a;
+
+DROP TABLE t1_local;
+DROP TABLE t2_local;
+DROP TABLE t1_all;
+DROP TABLE t2_all;
+DROP DATABASE test_02115;
```
Comment (vdimir) [inline tests/queries/0_stateless/02115_rewrite_local_join_right_distribute_table.sql:32]:
```suggestion
DROP DATABASE test_02115;

```

**Review by vdimir — APPROVED** [review]

File: tests/queries/0_stateless/02115_rewrite_local_join_right_distribute_table.sql [hunk 4]
```diff
@@ -0,0 +1,33 @@
+-- Tags: global, no-parallel
+CREATE DATABASE IF NOT EXISTS test_02115;
```
Comment (azat) [inline tests/queries/0_stateless/02115_rewrite_local_join_right_distribute_table.sql:2]: External database is not required for this tests, you can use `currentDatabase()` for `Distributed` instead. And then you will not need any tags for this tests.

**Pull request merged**